### DISCUSSION TITLE
insert_htmlbox with chinese/asian char

### DISCUSSION THREAD

**Discussion by tenghian** [pymupdf/PyMuPDF#4023] [General]
**Environment:**
Windows 10 (16GB): 7GB used
Python 3.10 + pymupdf 1.24.6
Python 3.12 + pymupdf 1.24.6
Python 3.10 + pymupdf 1.24.13
Python 3.12 + pymupdf 1.24.13

Code:
1. Created 5 blank pages with doc.new_page, inserted 230 paragraphs of chinese/Asian characters using insert_htmlbox().
-Saved with save(): PDF size is 592MB
-Saved with save(garbage=4, deflate=True): PDF size is 1.95MB

2. Created 5 blank pages with doc.new_page, used font=china-ss + doc.subset_fonts(), inserted 230 paragraphs of text with insert_htmlbox().
-Saved with save(): PDF size is 609MB
-Saved with save(garbage=4, deflate=True): PDF size is 250MB

Results:
-Foxit can open the PDF without issues.
-Adobe Acrobat 2024 crashes on both 609MB and 250KB files (though it works fine when each page is saved as a separate PDF or when subset_fonts is used for two pages only).

may i know How to modify the code to prevent Adobe Acrobat from crashing?

A 600MB uncompressed file for just 5 pages of chinese/Asian characters is excessive. For a 200-page PDF, uncompressed, it would require around 24GB of memory—unrealistic.

Additionally, converting the PDF to DOCX and back to PDF reduces the file size to around 350KB, which Adobe Acrobat can handle without issue

```
import pymupdf

rect = pymupdf.Rect(0, 0, 595, 842)
doc = pymupdf.open()

#page 1 =========================================================================================
page = doc.new_page(width=rect.width, height=rect.height)

css="* {font-family: cjk, sans-serif;font-size:18.002399444580078px;color:#221e1f}"
text="<b>年度报告</b>  <b> 2023</b>"
bbox=pymupdf.Rect(384.7019958496094, 587.5606689453125, 552.7543334960938, 609.01953125)
page.insert_htmlbox(bbox, text, css=css)

css="* {font-family: cjk, sans-serif;font-size:15.275199890136719px;color:#221e1f}"
text='<b>THREE-A RESOURCES BERHAD</b><span style="font-size:7.637599945068359;">注册号 199901006659 (481559-M)</span>'
bbox=pymupdf.Rect(42.448699951171875, 774.2913208007812, 280.0849609375, 801.104248046875)
page.insert_htmlbox(bbox, text, css=css)
```

[test44.zip](https://github.com/user-attachments/files/17627235/test44.zip)

**Comment by JorjMcKie** [comment]
I tried your scripts.
**_Not_** using the best possible compression options is not worth even looking at!
Why would anyone expect acceptable file sizes without this?
Method `page.insert_htmlbox` works based on `Story` objects. Font support here is under complete control of MuPDF's dynamic font selection.
This is even more true when one or more Asian (especially Chinese and Japanese) fonts come into play. The standard font supplied via MuPDF has a size of at least 3.6 MB (up to 50 MB depending on the selected MuPDF generation options).  This font is pulled in **_every time_** when using `insert_htmlbox`.
Therefore the documentation strongly recommends to use the best possible compression (`doc.ez_save(..., garbage=4)`) **_plus_** also building font subsets.

When doing this, your scripts delivers file sizes of 165 to 184 KB. Both work flawlessly under Adobe Acrobat as well.
Tested on Windows 11, Python 3.11, PyMuPDF 1.24.13.

**Comment by tenghian** [comment]
Ok thanks a lot for clarifying. ^_^
